File: 623.在二叉树中增加一行.cpp
/*
 * @lc app=leetcode.cn id=623 lang=cpp
 *
 * [623] 在二叉树中增加一行
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode *addOneRow(TreeNode *root, int val, int depth) {
        if (depth == 1) {
            return new TreeNode(val, root, nullptr);
        }
        int curDepth = 1;
        queue<TreeNode *> qu;
        qu.emplace(root);
        while (!qu.empty()) {
            int sz = qu.size();
            while (sz-- > 0) {
                TreeNode *curNode = qu.front();
                qu.pop();
                if (curDepth == depth - 1) {
                    TreeNode *a = new TreeNode(val, curNode->left, nullptr);
                    TreeNode *b = new TreeNode(val, nullptr, curNode->right);
                    curNode->left = a;
                    curNode->right = b;
                } else {
                    if (curNode->left != nullptr) qu.emplace(curNode->left);
                    if (curNode->right != nullptr) qu.emplace(curNode->right);
                }
            }
            ++curDepth;
        }
        return root;
    }
};
// @lc code=end

File: 623.在二叉树中增加一行.java
/*
 * @lc app=leetcode.cn id=623 lang=java
 *
 * [623] 在二叉树中增加一行
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if (depth == 1) {
            return new TreeNode(val, root, null);
        }
        int curDepth = 1;
        Deque<TreeNode> dq = new ArrayDeque<>();
        dq.addLast(root);
        while (!dq.isEmpty()) {
            int sz = dq.size();
            while (sz-- > 0) {
                TreeNode cur = dq.pollFirst();
                if (curDepth == depth - 1) {
                    TreeNode a = new TreeNode(val, cur.left, null);
                    TreeNode b = new TreeNode(val, null, cur.right);
                    cur.left = a;
                    cur.right = b;
                } else {
                    if (cur.left != null)
                        dq.addLast(cur.left);
                    if (cur.right != null)
                        dq.addLast(cur.right);
                }
            }
            ++curDepth;
        }
        return root;
    }
}
// @lc code=end
